### ISSUE TITLE
Chests

### ISSUE THREAD

**Issue by CraftYourBukkit** [Nukkit/Nukkit#1782]
I found a bug with chests. Sometimes I can't take stuff from chest.
I tested it with Windows 10 (not tested on mobile). There are no errors.

If I take a Item from a chest to my inventory the item gets invisible. But if I reopen the chest the item is back in it. 

I have no idea how to fix it

**Comment by LucarioDev** [comment]
Yeah I have the same bug, but only if I take the item too fast.

**Comment by CraftYourBukkit** [comment]
For me it's only for random chests. And if I can't take a item from chest1 it could be that another player can take one from it

**Comment by ScraM-Team** [comment]
Are these chests placed using the client?  I am having issues with chests created by a plugin, but I think it is a different issue.

**Comment by CraftYourBukkit** [comment]
Yes! Thanks for saying that! I tested it. It only happens with chests placed by a plugin

**Comment by ScraM-Team** [comment]
Is placing chests and setting their inventory documented?  I have fumbled around and come up with this, but I'm not sure it is the way I am supposed to use the API:

> // Minimal example, needs error checking in a real plugin
> level.setBlock(location, Block.CHEST);
> BlockEntityChest chest = (BlockEntityChest) level.getBlockEntity(location);
> chest.getInventory().setItem(slot, Item.get(Item.APPLE, 0, 4));

**Comment by CraftYourBukkit** [comment]
I have to add something:

The chests are placed manually but the world is copied. But the main reason I think is this:
The items in the chest are placed by a plugin

**Comment by ScraM-Team** [comment]
Are you adding items in the chest using code like I showed above?  If you have a different way, please share.  I haven't seen a Nukkit chest sample plugin.

**Comment by CraftYourBukkit** [comment]
Yes I do it like your example

**Comment by CraftYourBukkit** [comment]
Is there a way to fix this?

**Comment by ScraM-Team** [comment]
@CraftYourBukkit , should we take a crack at it?  Have you built Nukkit from source?  

@Snake1999 , any advice on where to start?

**Comment by CraftYourBukkit** [comment]
I got Nukkit from Jenkins (newest version). And I hope u will fix it

**Comment by ScraM-Team** [comment]
I am still looking at this.  No Nukkit changes yet, however.  I am still confirming the behavior and reliable test cases.  I will report more as I find it.

**Comment by ScraM-Team** [comment]
One related factor for me is I'm loading a map that has chests already on it, then trying to place chests at those same locations.  It seems to help sometimes to place air where the map chests were, then place new chests, then fill them.  Otherwise the chests show up invisible or never get filled with items.

I also see issues with grabbing items out of the chest, which is the original issue described here.  You have to tap pretty fast, though, to make it happen (at least for me).  When I tap one item after the other, quickly, not all of them actually show up in my inventory even though the animation shows them flying in.  It looks like a timing/race condition, since if I tap more slowly, it doesn't happen.

**Comment by CraftYourBukkit** [comment]
The tapping speed doesn't change anything for me (win10)

**Comment by ScraM-Team** [comment]
It seems you have to create the Items NBT tag after creating the block.  In my opinion this should be handled by the API, but it isn't yet when doing setBlock.  In the meanwhile, do this when creating a chest:

> // Minimal example, needs error checking in a real plugin
> level.setBlock(location, Block.CHEST);
> Block block = level.getBlock(position);
> CompoundTag nbt = new CompoundTag("")
>                .putList(new ListTag<>("Items"))
>                .putString("id", BlockEntity.CHEST)
>                .putInt("x", (int) block.x)
>                .putInt("y", (int) block.y)
>                .putInt("z", (int) block.z);
> BlockEntityChest chest = new BlockEntityChest(this.getLevel().getChunk((int) (block.x) >> 4, (int) (block.z) >> 4), nbt);
> level.addBlockEntity(chest);
> chest.getInventory().setItem(slot, Item.get(Item.APPLE, 0, 4));

To Nukkit devs, I see code to do this in BlockChest.place() and .onActivate().  Could I refactor this out into a separate method?  Could this be invoked automatically during Level.setBlock()?
